src/locker.rs

```diff
@@ -0,0 +1,112 @@
+use std::ops::{Deref, DerefMut};
+
+use crate::isolate::Isolate;
+use crate::scope::data::ScopeData;
+use crate::support::Opaque;
+
+#[repr(C)]
+#[derive(Debug)]
+struct LockerHandle(Opaque);
+
+/// A handle to a shared isolate, allowing access to the isolate in a thread safe way.
+///
+/// Unlike V8 isolates, these do not currently support re-entrancy.
+/// Do not create multiple lockers to the same isolate in the same thread.
+#[derive(Debug)]
+pub struct Locker<'a> {
+  _lock: raw::Locker,
+  // We maintain a mut reference to ensure we have exclusive ownership of the isolate during the lock.
+  locked: &'a mut Isolate,
+}
+
+impl<'a> Locker<'a> {
+  /// Claims the isolate, this should only be used from a shared isolate.
+  pub(crate) fn new(isolate: &'a mut Isolate) -> Self {
```

Comment:
i think this is UB, the locker should be acquired before a mutable reference is created.

Reply:
Update: set parameter to `isolate: &Isolate` (while unsafe cast to `&mut Isolate` in this `new` function).